cjwstate/modules/i18n/catalogs/extract/code.py

```diff
@@ -0,0 +1,212 @@
+from babel.messages.extract import extract
+import pathlib
+import re
+from typing import Dict, Any
+from babel.util import parse_encoding, parse_future_flags
+from babel._compat import PY2, text_type
+from tokenize import generate_tokens, COMMENT, NAME, OP, STRING
+
+
+_default_message_re = re.compile(r"\s*default-message:\s*(.*)\s*")
+
+
+def find_messages_in_module_code(
+    code_path: pathlib.Path, root_path: pathlib.Path
+) -> Dict[str, Dict[str, Any]]:
+    with open(code_path, "rb") as code_file:
+        return _find_messages_in_module_code(
+            code_file, str(code_path.relative_to(root_path))
+        )
+
+
+def _find_messages_in_module_code(
+    code_file, relative_path_name: str
+) -> Dict[str, Dict[str, Any]]:
+    messages_data = extract(
+        _extract_module_code,
+        code_file,
+        keywords={"trans": (1, 2)},
+        comment_tags=["i18n"],
+    )
+    messages = {}
+    for lineno, message_id, comments, context in messages_data:
+        default_message = ""
+        for comment in comments:
+            match = _default_message_re.match(comment)
+            if match:
+                default_message = match.group(1).strip()
+                comments.remove(comment)
+        if message_id in messages:
+            messages[message_id]["comments"].extend(comments)
+            messages[message_id]["locations"].append((relative_path_name, lineno))
+        else:
+            messages[message_id] = {
+                "string": default_message,
+                "comments": comments,
+                "locations": [(relative_path_name, lineno)],
+            }
+    return messages
+
+
+def _extract_module_code(fileobj, keywords, comment_tags, options):
```

Comment:
How is this different from `cjworkbench.i18n.catalogs.extract.extract_python()`?

Should the two functions be merged? It seems to me they should, but you may be thinking of something I'm missing.

Reply:
Indeed, most of the code was the same (parsing logic copied by `pybabel`). I changed the structure a bit, in order to avoid code duplication (and at the same time highlight the difference between the two): now the common part expects a callback to run when a translation function has been found.